File: backend/entity/Topic.go
package entity

import (
	"gorm.io/gorm"
)

type Topic struct {
    gorm.Model
    Title         string `json:"title" form:"title" valid:"required~Title is required"`
    Objective     string `json:"objective" form:"objective" valid:"required~Objective is required"`
    Scope         string `json:"scope" form:"scope" valid:"required~Scope is required"`
    Description   string `json:"description" form:"description" valid:"required~Description is required"`
	Status 	  string `json:"status" form:"status" valid:"required~Status is required"`
    FileAttachment string `json:"file_attachment,omitempty" form:"file_attachment"`
    ProposerRole string `json:"proposer_role" form:"proposer_role"`

	GroupProjectID *uint `json:"group_project_id" form:"group_project_id"`
	GroupProject   *GroupProject `gorm:"foreignKey:GroupProjectID" json:"group_project"`
	
	TeacherID *uint `json:"teacher_id" form:"teacher_id"`
	Teacher   *User `gorm:"foreignKey:TeacherID" json:"teacher"`
	
    TopicApproval []TopicApproval `gorm:"foreignKey:TopicID" json:"topic_approvals"`

}

File: frontend/app/interfaces/Topic.ts
import { GroupProject } from './Group';

export type TopicStatus = 'Approved' | 'Rejected' | 'Pending' | 'Closed';

export const statusMap: Record<TopicStatus, { color: string; text: string }> = {
  Approved: { color: 'success', text: 'อนุมัติแล้ว' },
  Rejected: { color: 'error', text: 'ไม่อนุมัติ' },
  Pending:  { color: 'warning', text: 'รอพิจารณา' },
  Closed:   { color: 'default', text: 'ปิดหัวข้อแล้ว' },
};


export interface Topic {
    ID: number;
    id?: number;
    title: string;
    objective: string;
    scope: string;
    description: string;
    status: TopicStatus;
    file_attachment?: string;
    proposer_role: string;
    teacher_id?: number;
    group_project_id?: number;

    // Relations
    topic_approvals?: TopicApproval[];
    group_project?: GroupProject;
}

export interface TopicApproval {
    ID: number;
    topic_id: number;
    status: TopicStatus;
    comment: string;
    teacher_id: number;
    approval_date?: string;
}